### ISSUE TITLE
CLIP does not support all OOD methods

### ISSUE THREAD

**Issue by saibr** [Jingkang50/OpenOOD#195]
Hello! Thank you for creating this repository. 

I noticed that currently the use of CLIP is only supported for some of the methods, since some functions like 'forward_threshold' are not well integrated. Could this be integrated in the already existing CLIPZeroshot class as follows?

```
class CLIPZeroshot(nn.Module):
    def __init__(self, classnames, templates, backbone='ViT-B/16'):
        super().__init__()
        assert backbone in clip.available_models()
        self.model, self.preprocess = clip.load(backbone, device='cuda')
        self.zeroshot_weights = zeroshot_classifier(self.model, classnames,
                                                    templates)

    def forward(self, x, return_feature=False):
        image_features = self.model.encode_image(x)
        image_features /= image_features.norm(dim=-1, keepdim=True)
        logits = image_features @ self.zeroshot_weights
        if return_feature:
            return logits, image_features
        else:
            return logits

    def forward_threshold(self, x, threshold):
        image_features = self.model.encode_image(x)
        image_features /= image_features.norm(dim=-1, keepdim=True)
        image_features = image_features.clip(max=threshold)
        logits = image_features @ self.zeroshot_weights
        return logits
```

I'm not sure if it can be done like this or if it's more complicated, because we also have a second text modality. Any help would be appreciated. Thank you!

**Comment by zjysteven** [comment]
Hi @saibr, yes I don't see problems with doing so, and adapting existing methods that leverage penultimate features will be worth trying for sure.

The reason we were only supporting methods that use the logits in is that, in an updated manuscript (which we haven't released), we were specifically investigating the zero-shot OOD detection performance of CLIP. To my knowledge, methods that use penultimate features will break the "zero-shot" assumption (e.g., MDS and KNN will track the features of each class's training samples). And in such cases, fine-tuning the image encoder with the available training images could potentially lead to better performance.
